File: src/features/shopping/budget.test.ts
import { describe, expect, it } from 'vitest'
import { calculateBudget, itemEffectivePrice } from '@/features/shopping/budget'
import type { ShoppingItem } from '@/types/models'

function makeItem(overrides: Partial<ShoppingItem>): ShoppingItem {
  return {
    id: crypto.randomUUID(),
    name: 'פריט',
    category: 'תינוקייה',
    priority: 'normal',
    status: 'need',
    alternatives: [],
    createdAt: 0,
    ...overrides,
  }
}

describe('itemEffectivePrice', () => {
  it('prefers actualPrice over everything else', () => {
    const item = makeItem({ actualPrice: 100, estimatedPrice: 50 })
    expect(itemEffectivePrice(item)).toBe(100)
  })

  it('falls back to the chosen alternative price when no actualPrice is set', () => {
    const item = makeItem({
      estimatedPrice: 50,
      chosenAlternativeId: 'alt-1',
      alternatives: [{ id: 'alt-1', name: 'חלופה', price: 80 }],
    })
    expect(itemEffectivePrice(item)).toBe(80)
  })

  it('falls back to estimatedPrice when nothing else is set', () => {
    const item = makeItem({ estimatedPrice: 50 })
    expect(itemEffectivePrice(item)).toBe(50)
  })

  it('is undefined when no price is known', () => {
    expect(itemEffectivePrice(makeItem({}))).toBeUndefined()
  })
})

describe('calculateBudget', () => {
  it('only counts spent for bought items', () => {
    const items = [
      makeItem({ status: 'bought', actualPrice: 100 }),
      makeItem({ status: 'need', estimatedPrice: 40 }),
    ]
    const totals = calculateBudget(items)
    expect(totals.spentTotal).toBe(100)
    expect(totals.estimatedTotal).toBe(140)
    expect(totals.boughtCount).toBe(1)
    expect(totals.totalCount).toBe(2)
  })

  it('groups totals by category', () => {
    const items = [
      makeItem({ category: 'בגדים', estimatedPrice: 20 }),
      makeItem({ category: 'בגדים', estimatedPrice: 30 }),
      makeItem({ category: 'האכלה', estimatedPrice: 10 }),
    ]
    const totals = calculateBudget(items)
    const clothes = totals.byCategory.find((c) => c.category === 'בגדים')
    expect(clothes?.estimated).toBe(50)
  })

  it('handles an empty list', () => {
    const totals = calculateBudget([])
    expect(totals).toEqual({
      estimatedTotal: 0,
      spentTotal: 0,
      boughtCount: 0,
      totalCount: 0,
      byCategory: [],
    })
  })
})

File: android/core/domain/src/test/kotlin/com/oryareach/core/domain/shopping/BudgetTest.kt
package com.oryareach.core.domain.shopping

import com.oryareach.core.model.Priority
import com.oryareach.core.model.ShoppingAlternative
import com.oryareach.core.model.ShoppingCategory
import com.oryareach.core.model.ShoppingItem
import com.oryareach.core.model.ShoppingStatus
import io.kotest.matchers.shouldBe
import org.junit.Test

private fun item(
    category: ShoppingCategory = ShoppingCategory.NURSERY,
    estimatedPrice: Int? = null,
    actualPrice: Int? = null,
    status: ShoppingStatus = ShoppingStatus.NEED,
    alternatives: List<ShoppingAlternative> = emptyList(),
    chosenAlternativeId: String? = null,
) = ShoppingItem(
    id = "id",
    name = "item",
    category = category,
    estimatedPrice = estimatedPrice,
    actualPrice = actualPrice,
    priority = Priority.NORMAL,
    status = status,
    alternatives = alternatives,
    chosenAlternativeId = chosenAlternativeId,
)

class BudgetTest {

    @Test
    fun `itemEffectivePrice prefers actualPrice over everything else`() {
        itemEffectivePrice(item(actualPrice = 100, estimatedPrice = 50)) shouldBe 100
    }

    @Test
    fun `itemEffectivePrice falls back to the chosen alternative price`() {
        val i = item(
            estimatedPrice = 50,
            chosenAlternativeId = "alt-1",
            alternatives = listOf(ShoppingAlternative(id = "alt-1", name = "alt", price = 80)),
        )
        itemEffectivePrice(i) shouldBe 80
    }

    @Test
    fun `itemEffectivePrice falls back to estimatedPrice when nothing else is set`() {
        itemEffectivePrice(item(estimatedPrice = 50)) shouldBe 50
    }

    @Test
    fun `itemEffectivePrice is null when no price is known`() {
        itemEffectivePrice(item()) shouldBe null
    }

    @Test
    fun `calculateBudget only counts spent for bought items`() {
        val totals = calculateBudget(
            listOf(
                item(status = ShoppingStatus.BOUGHT, actualPrice = 100),
                item(status = ShoppingStatus.NEED, estimatedPrice = 40),
            ),
        )
        totals.spentTotal shouldBe 100
        totals.estimatedTotal shouldBe 140
        totals.boughtCount shouldBe 1
        totals.totalCount shouldBe 2
    }

    @Test
    fun `calculateBudget groups totals by category`() {
        val totals = calculateBudget(
            listOf(
                item(category = ShoppingCategory.CLOTHING, estimatedPrice = 20),
                item(category = ShoppingCategory.CLOTHING, estimatedPrice = 30),
                item(category = ShoppingCategory.FEEDING, estimatedPrice = 10),
            ),
        )
        totals.byCategory.first { it.category == ShoppingCategory.CLOTHING }.estimated shouldBe 50
    }

    @Test
    fun `calculateBudget handles an empty list`() {
        val totals = calculateBudget(emptyList())
        totals.estimatedTotal shouldBe 0
        totals.spentTotal shouldBe 0
        totals.boughtCount shouldBe 0
        totals.totalCount shouldBe 0
        totals.byCategory shouldBe emptyList()
    }
}
